applications/virtual-fly-brain/client/src/components/ThreeDCanvas.js

```diff
@@ -55,213 +53,57 @@ class ThreeDCanvas extends Component {
-
-    newData?.forEach( dat => {
-      dat.visible = data?.find( i => i.Id === data.instancePath )?.visible;
-    })
-    let mappedCanvasData = [...newData]
-    let match = mappedCanvasData?.find( m => instance.id === m.instancePath )
-    if ( match ){
-      match.visible = true;
-      this?.canvasRef?.current?.threeDEngine?.updateInstances(mappedCanvasData)
-      this.setState({ ...this.state, mappedCanvasData : mappedCanvasData})
-    }
-  }
-
-  componentDidUpdate(prevProps, prevState) {
+  componentDidUpdate(prevProps) {
     if(this.props.event.trigger !== prevProps.event.trigger){ 
-      let that = this;
-      let allLoadedInstances = this.props.allLoadedInstances;
-      const mappedCanvasData = [...this.state.mappedCanvasData];
-      const targetInstance = allLoadedInstances?.find( i => i.metadata?.Id === this.props.event.id)
-      const focusInstance = window.Instances?.find( instance => instance.wrappedObj.id === this.props.focusInstance?.Id);
+      const mappedCanvasData = [...this.props.allLoadedInstances];
+      const threeDObjects = [...this.props.threeDObjects];
 
       switch(this.props.event.action){
         case getInstancesTypes.FOCUS_INSTANCE:
-          if ( focusInstance){
-            this.canvasRef.current.threeDEngine.cameraManager.zoomTo([focusInstance])
-          } else {
-            this.canvasRef.current.defaultCameraControlsHandler("cameraHome")
-          }
           break;
-        case getInstancesTypes.CHANGE_COLOR:
-          this.updateColors(targetInstance,mappedCanvasData)
+        case getInstancesTypes.ADD_INSTANCE:{
+          const instance = mappedCanvasData.find( i => i.metadata?.Id === this.props.event.id );
+          if ( instance?.simpleInstance === undefined ) {
+            get3DMesh(instance)
+          } 
           break;
-        case getInstancesTypes.ADD_INSTANCE:
-          // Set all existing instances to invisible
-          if ( mappedCanvasData?.find( i => targetInstance?.metadata?.Id === i.instancePath ) === undefined ){
-            if (targetInstance?.metadata?.Images)
-            {
-              fetch(targetInstance?.metadata?.Images?.[Object.keys(targetInstance?.metadata?.Images)[0]][0].obj)
-              .then(response => response.text())
-              .then(base64Content => {
-                const instance = {
-                  "eClass": "SimpleInstance",
-                  "id": targetInstance?.metadata?.Id,
-                  "name": targetInstance?.metadata?.Name,
-                  "type": { "eClass": "SimpleType" },
-                  "visualValue": {
-                    "eClass": Resources.OBJ,
-                    'obj': base64Content
-                  }, 
-                  "visible" : true,
-                  "color" : targetInstance?.color
-                }
-                that.newInstance(instance);
-              });
+        }
+        case getInstancesTypes.UPDATE_INSTANCES:{
```

Comment:
again, this should happen in the reducer imho

Reply:
@ddelpiano get3DMesh() is a redux action call that happens when the 3D Mesh is needed. Calling it from the reducer will be anti pattern. The idea behind this, is to let the Canvas request the 3D Mesh when it needs it.

Reply:
This action [happens in the reducer ](https://github.com/MetaCell/virtual-fly-brain/pull/97/files/5341ad31e7d135a59b3368c461fdaf9734c50da7#diff-97c2e04c182ee29a695079add533ec8937a02319e87db6185fc3d918fdc3ce8eR81). In here, what we are doing it's formatting the Instances in the reducer state into a format that the canvas take. I'm not sure we need the redux store saving the mappedCanvasData.